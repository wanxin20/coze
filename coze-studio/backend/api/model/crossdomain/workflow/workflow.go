/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package workflow

type Locator uint8

const (
	FromDraft Locator = iota
	FromSpecificVersion
	FromLatestVersion
)

type ExecuteConfig struct {
	ID            int64
	From          Locator
	Version       string
	CommitID      string
	Operator      int64
	Mode          ExecuteMode
	AppID         *int64
	AgentID       *int64
	ConnectorID   int64
	ConnectorUID  string
	TaskType      TaskType
	SyncPattern   SyncPattern
	InputFailFast bool // whether to fail fast if input conversion has warnings
	BizType       BizType
	Cancellable   bool
}

type ExecuteMode string

const (
	ExecuteModeDebug     ExecuteMode = "debug"
	ExecuteModeRelease   ExecuteMode = "release"
	ExecuteModeNodeDebug ExecuteMode = "node_debug"
)

type TaskType string

const (
	TaskTypeForeground TaskType = "foreground"
	TaskTypeBackground TaskType = "background"
)

type SyncPattern string

const (
	SyncPatternSync   SyncPattern = "sync"
	SyncPatternAsync  SyncPattern = "async"
	SyncPatternStream SyncPattern = "stream"
)

var DebugURLTpl = "http://127.0.0.1:3000/work_flow?execute_id=%d&space_id=%d&workflow_id=%d&execute_mode=2"

type BizType string

const (
	BizTypeAgent    BizType = "agent"
	BizTypeWorkflow BizType = "workflow"
)
