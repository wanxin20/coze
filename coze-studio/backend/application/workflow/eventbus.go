/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package workflow

import (
	"context"

	"github.com/coze-dev/coze-studio/backend/api/model/resource/common"
	"github.com/coze-dev/coze-studio/backend/domain/search/entity"
	search "github.com/coze-dev/coze-studio/backend/domain/search/entity"
	"github.com/coze-dev/coze-studio/backend/domain/search/service"
)

var eventBus service.ResourceEventBus

func setEventBus(bus service.ResourceEventBus) {
	eventBus = bus
}

func PublishWorkflowResource(ctx context.Context, workflowID int64, mode *int32, op search.OpType, r *search.ResourceDocument) error {
	if r == nil {
		r = &search.ResourceDocument{}
	}

	r.ResType = common.ResType_Workflow
	r.ResID = workflowID
	r.ResSubType = mode

	event := &entity.ResourceDomainEvent{
		OpType:   entity.OpType(op),
		Resource: r,
	}

	if op == search.Created {
		event.Resource.CreateTimeMS = r.CreateTimeMS
		event.Resource.UpdateTimeMS = r.UpdateTimeMS
	} else if op == search.Updated {
		event.Resource.UpdateTimeMS = r.UpdateTimeMS
	}

	err := eventBus.PublishResources(ctx, event)
	if err != nil {
		return err
	}

	return nil
}
