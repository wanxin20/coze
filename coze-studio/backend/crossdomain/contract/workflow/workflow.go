/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package crossworkflow

import (
	"context"

	"github.com/cloudwego/eino/compose"
	einoCompose "github.com/cloudwego/eino/compose"
	"github.com/cloudwego/eino/schema"

	workflowModel "github.com/coze-dev/coze-studio/backend/api/model/crossdomain/workflow"
	"github.com/coze-dev/coze-studio/backend/domain/workflow"
	"github.com/coze-dev/coze-studio/backend/domain/workflow/entity"
	workflowEntity "github.com/coze-dev/coze-studio/backend/domain/workflow/entity"

	"github.com/coze-dev/coze-studio/backend/domain/workflow/entity/vo"
)

// TODO (@fanlv): Parameter references need to be modified.
type Workflow interface {
	WorkflowAsModelTool(ctx context.Context, policies []*vo.GetPolicy) ([]workflow.ToolFromWorkflow, error)
	WithResumeToolWorkflow(resumingEvent *workflowEntity.ToolInterruptEvent, resumeData string,
		allInterruptEvents map[string]*workflowEntity.ToolInterruptEvent) einoCompose.Option
	ReleaseApplicationWorkflows(ctx context.Context, appID int64, config *ReleaseWorkflowConfig) ([]*vo.ValidateIssue, error)
	GetWorkflowIDsByAppID(ctx context.Context, appID int64) ([]int64, error)
	SyncExecuteWorkflow(ctx context.Context, config workflowModel.ExecuteConfig, input map[string]any) (*workflowEntity.WorkflowExecution, vo.TerminatePlan, error)
	WithExecuteConfig(cfg workflowModel.ExecuteConfig) einoCompose.Option
	WithMessagePipe() (compose.Option, *schema.StreamReader[*entity.Message])
}

type ExecuteConfig = workflowModel.ExecuteConfig
type ExecuteMode = workflowModel.ExecuteMode
type NodeType = entity.NodeType

type WorkflowMessage = entity.Message

const (
	NodeTypeOutputEmitter NodeType = "OutputEmitter"
)

const (
	ExecuteModeDebug     ExecuteMode = "debug"
	ExecuteModeRelease   ExecuteMode = "release"
	ExecuteModeNodeDebug ExecuteMode = "node_debug"
)

type TaskType = workflowModel.TaskType

const (
	TaskTypeForeground TaskType = "foreground"
	TaskTypeBackground TaskType = "background"
)

type BizType = workflowModel.BizType

const (
	BizTypeAgent    BizType = "agent"
	BizTypeWorkflow BizType = "workflow"
)

type ReleaseWorkflowConfig = vo.ReleaseWorkflowConfig

type ToolInterruptEvent = workflowEntity.ToolInterruptEvent

var defaultSVC Workflow

func DefaultSVC() Workflow {
	return defaultSVC
}

func SetDefaultSVC(svc Workflow) {
	defaultSVC = svc
}
