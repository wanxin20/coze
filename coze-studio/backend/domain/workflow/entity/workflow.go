/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package entity

import (
	"github.com/coze-dev/coze-studio/backend/api/model/workflow"
	"github.com/coze-dev/coze-studio/backend/domain/workflow/entity/vo"
)

type ContentType = workflow.WorkFlowType
type Tag = workflow.Tag
type Mode = workflow.WorkflowMode

type Workflow struct {
	ID       int64
	CommitID string

	*vo.Meta
	*vo.CanvasInfo
	*vo.DraftMeta
	*vo.VersionMeta
}

func (w *Workflow) GetBasic() *WorkflowBasic {
	var version string
	if w.VersionMeta != nil {
		version = w.VersionMeta.Version
	}
	return &WorkflowBasic{
		ID:       w.ID,
		Version:  version,
		SpaceID:  w.SpaceID,
		AppID:    w.AppID,
		CommitID: w.CommitID,
	}
}

func (w *Workflow) GetLatestVersion() string {
	if w.LatestPublishedVersion == nil {
		return ""
	}

	return *w.LatestPublishedVersion
}

func (w *Workflow) GetVersion() string {
	if w.VersionMeta == nil {
		return ""
	}
	return w.VersionMeta.Version
}

type IDVersionPair struct {
	ID      int64
	Version string
}

type WorkflowBasic struct {
	ID       int64
	Version  string
	SpaceID  int64
	AppID    *int64
	CommitID string
}

type CopyWorkflowFromAppToLibraryResult struct {
	WorkflowIDVersionMap map[int64]IDVersionPair
	ValidateIssues       []*vo.ValidateIssue
	CopiedWorkflows      []*Workflow
}
