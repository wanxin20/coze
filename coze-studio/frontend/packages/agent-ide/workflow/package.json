{
  "name": "@coze-agent-ide/workflow",
  "version": "0.0.1",
  "description": "bot 与 workflow 交互的内容",
  "license": "Apache-2.0",
  "author": "meixuliang.3@bytedance.com",
  "maintainers": [],
  "main": "src/index.tsx",
  "scripts": {
    "build": "exit 0",
    "lint": "eslint ./ --cache",
    "test": "vitest --run --passWithNoTests",
    "test:cov": "npm run test -- --coverage"
  },
  "dependencies": {
    "@coze-agent-ide/bot-plugin-export": "workspace:*",
    "@coze-agent-ide/navigate": "workspace:*",
    "@coze-agent-ide/plugin-shared": "workspace:*",
    "@coze-agent-ide/space-bot": "workspace:*",
    "@coze-agent-ide/tool": "workspace:*",
    "@coze-agent-ide/workflow-as-agent-adapter": "workspace:*",
    "@coze-arch/bot-api": "workspace:*",
    "@coze-arch/bot-error": "workspace:*",
    "@coze-arch/bot-flags": "workspace:*",
    "@coze-arch/bot-hooks": "workspace:*",
    "@coze-arch/bot-icons": "workspace:*",
    "@coze-arch/bot-semi": "workspace:*",
    "@coze-arch/bot-studio-store": "workspace:*",
    "@coze-arch/bot-utils": "workspace:*",
    "@coze-arch/coze-design": "0.0.6-alpha.346d77",
    "@coze-arch/i18n": "workspace:*",
    "@coze-arch/idl": "workspace:*",
    "@coze-arch/report-events": "workspace:*",
    "@coze-arch/report-tti": "workspace:*",
    "@coze-common/assets": "workspace:*",
    "@coze-studio/bot-detail-store": "workspace:*",
    "@coze-studio/components": "workspace:*",
    "@coze-workflow/base": "workspace:*",
    "@coze-workflow/components": "workspace:*",
    "ahooks": "^3.7.8",
    "classnames": "^2.3.2",
    "copy-to-clipboard": "^3.3.3",
    "immer": "^10.0.3",
    "lodash-es": "^4.17.21",
    "zustand": "^4.4.7"
  },
  "devDependencies": {
    "@coze-arch/bot-env": "workspace:*",
    "@coze-arch/bot-typings": "workspace:*",
    "@coze-arch/eslint-config": "workspace:*",
    "@coze-arch/stylelint-config": "workspace:*",
    "@coze-arch/ts-config": "workspace:*",
    "@coze-arch/vitest-config": "workspace:*",
    "@testing-library/jest-dom": "^6.1.5",
    "@testing-library/react": "^14.1.2",
    "@testing-library/react-hooks": "^8.0.1",
    "@types/lodash-es": "^4.17.10",
    "@types/react": "18.2.37",
    "@types/react-dom": "18.2.15",
    "@vitest/coverage-v8": "~3.0.5",
    "react": "~18.2.0",
    "react-dom": "~18.2.0",
    "stylelint": "^15.11.0",
    "vite-plugin-svgr": "~3.3.0",
    "vitest": "~3.0.5"
  },
  "peerDependencies": {
    "react": ">=18.2.0",
    "react-dom": ">=18.2.0"
  },
  "// deps": "immer@^10.0.3 为脚本自动补齐，请勿改动"
}

