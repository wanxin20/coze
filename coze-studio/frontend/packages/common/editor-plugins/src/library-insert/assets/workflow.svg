<svg width="15" height="14" viewBox="0 0 15 14" fill="none" xmlns="http://www.w3.org/2000/svg">
<g id="Component 2">
<path id="Union" fill-rule="evenodd" clip-rule="evenodd" d="M9.06505 5.38087C9.27846 5.65615 9.61235 5.83333 9.98763 5.83333H12.6126C13.257 5.83333 13.7793 5.311 13.7793 4.66667V2.04167C13.7793 1.39733 13.257 0.875 12.6126 0.875H9.98763C9.3433 0.875 8.82096 1.39733 8.82096 2.04167V4.19269L8.80691 4.16985L6.48763 5.5971V5.25C6.48763 4.60567 5.9653 4.08333 5.32096 4.08333L2.69596 4.08333C2.05163 4.08333 1.5293 4.60567 1.5293 5.25V8.75C1.5293 9.39433 2.05163 9.91667 2.69596 9.91667H5.32096C5.9653 9.91667 6.48763 9.39433 6.48763 8.75V8.36035L8.82096 9.72859V11.9583C8.82096 12.6027 9.3433 13.125 9.98763 13.125H12.6126C13.257 13.125 13.7793 12.6027 13.7793 11.9583V9.33333C13.7793 8.689 13.257 8.16667 12.6126 8.16667H9.98763C9.64375 8.16667 9.33463 8.31544 9.12111 8.55214L6.48763 7.00789V6.96697L9.06505 5.38087ZM9.98763 2.04167V4.66667H12.6126V2.04167H9.98763ZM9.98763 9.33333V11.9583H12.6126V9.33333H9.98763ZM2.69596 8.75L2.69596 5.25L5.32096 5.25L5.32096 8.75H2.69596Z" fill="#4E40E5"/>
</g>
</svg>
