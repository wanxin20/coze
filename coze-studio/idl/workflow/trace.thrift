include "../base.thrift"

namespace go workflow

enum FrontedTagType {
    TEXT = 0 // Text
    TIME = 1 // Time, with timestamp, in milliseconds
    TIME_DURATION = 2 // Time interval, in milliseconds
}

// Tag
struct TraceTag {
    1: string Key (go.tag = "json:\"key\"", agw.key = "key")
    2: TagType TagType (go.tag = "json:\"tag_type\"", agw.key = "tag_type")
    3: Value Value (go.tag = "json:\"value\"", agw.key = "value")
}

struct FilterTag {
    1: optional string DataType (go.tag = "json:\"data_type\"", agw.key = "data_type")
    2: optional string TagKey (go.tag = "json:\"tag_key\"", agw.key = "tag_key")
    3: optional list<string> MultiTagKeys (go.tag = "json:\"multi_tag_keys\"", agw.key = "multi_tag_keys")
    4: optional list<string> Values (go.tag = "json:\"values\"", agw.key = "values")
    5: optional QueryTypeEnum QueryType (go.tag = "json:\"query_type\"", agw.key = "query_type")
}

enum QueryTypeEnum {
    Undefined = 0
    Match = 1
    Term = 2
    Range = 3
    Exist = 4
    NotExist = 5
}

enum SpanStatus{
    Unknown = 0
    Success = 1
    Fail = 2
}

struct ListRootSpansRequest {
    2: required i64 StartAt (go.tag = "json:\"start_at\"", api.body = "start_at") // It's in milliseconds.
    3: required i64 EndAt (go.tag = "json:\"end_at\"", api.body = "end_at") // It's in milliseconds.
    4: optional i16 Limit (go.tag = "json:\"limit\"", api.body = "limit")
    5: optional bool DescByStartTime (go.tag = "json:\"desc_by_start_time\"", api.body = "desc_by_start_time")
    6: optional i32 Offset (go.tag = "json:\"offset\"", api.body = "offset")
    7: required string WorkflowID(go.tag = "json:\"workflow_id\"", api.body = "workflow_id")
    8: optional string Input(go.tag = "json:\"input\"", api.body = "input")
    9: optional SpanStatus Status(go.tag = "json:\"status\"", api.body = "status")
    10: optional i32 ExecuteMode (go.tag = "json:\"execute_mode\"", api.body = "execute_mode") // Formal run/practice run/Node Debug

    255: optional base.Base Base
}

struct Span {
    1: string TraceID (go.tag = "json:\"trace_id\"", agw.key = "trace_id")
    2: string LogID (go.tag = "json:\"log_id\"", agw.key = "log_id")
    3: string PSM (go.tag = "json:\"psm\"", agw.key = "psm")
    4: string DC (go.tag = "json:\"dc\"", agw.key = "dc")
    5: string PodName (go.tag = "json:\"pod_name\"", agw.key = "pod_name")
    6: string SpanID (go.tag = "json:\"span_id\"", agw.key = "span_id")
    7: string Type (go.tag = "json:\"type\"", agw.key = "type")
    8: string Name (go.tag = "json:\"name\"", agw.key = "name")
    9: string ParentID (go.tag = "json:\"parent_id\"", agw.key = "parent_id")
    10: i64 Duration (go.tag = "json:\"duration\"", agw.key = "duration") // It's in milliseconds.
    11: i64 StartTime (go.tag = "json:\"start_time\"", agw.key = "start_time") // It's in milliseconds.
    12: i32 StatusCode (go.tag = "json:\"status_code\"", agw.key = "status_code")
    13: list<TraceTag> Tags (go.tag = "json:\"tags\"", agw.key = "tags")
}

struct Value {
    1: optional string VStr (go.tag = "json:\"v_str\"", agw.key = "v_str")
    2: optional double VDouble (go.tag = "json:\"v_double\"", agw.key = "v_double")
    3: optional bool VBool (go.tag = "json:\"v_bool\"", agw.key = "v_bool")
    4: optional i64 VLong (go.tag = "json:\"v_long\"", agw.key = "v_long")
    5: optional binary VBytes (go.tag = "json:\"v_bytes\"", agw.key = "v_bytes")
}

enum TagType {
    STRING = 0
    DOUBLE = 1
    BOOL = 2
    LONG = 3
    BYTES = 4
}

struct ListRootSpansResponse {
    1: optional list<Span> Spans (go.tag = "json:\"spans\"", api.body = "spans")

    255: optional base.BaseResp BaseResp
}


struct GetTraceSDKRequest {
    2: optional string LogID (go.tag = "json:\"log_id\"", api.query = "log_id")
    4: optional i64 StartAt (go.tag = "json:\"start_at\"", api.query = "start_at") // It's in milliseconds.
    5: optional i64 EndAt (go.tag = "json:\"end_at\"", api.query = "end_at") // It's in milliseconds.
    6: optional i64 WorkflowID (go.tag = "json:\"workflow_id\"", api.query = "workflow_id")
    7: optional i64 ExecuteID (go.tag = "json:\"execute_id\"", api.query = "execute_id")

    255: optional base.Base Base
}

enum QueryScene {
    ALICE_OP = 0 // Doubao cici full link debugging station
    DOUBAO_CICI_DEBUG = 1 // Doubao cici debugging function
    WORKFLOW_DEBUG = 2 // Workflow debugging
}

enum TenantLevel {
    Ordinary = 0
    AdvancedWhitelist = 1
}

struct GetTraceSDKResponse {
    1: optional TraceFrontend data
    255: optional base.BaseResp BaseResp
}


struct KeyScene {
    1: optional string Scene (go.tag = "json:\"scene\"", agw.key = "scene") // Scenarios such as "Split search terms"\ "Search"
    2: optional string StatusMessage (go.tag = "json:\"status_message\"", agw.key = "status_message") // status information
    3: optional string System (go.tag = "json:\"system\"", agw.key = "system")
    4: optional list<MessageItem> HistoryMessages (go.tag = "json:\"history_messages\"", agw.key = "history_messages") // chat history
    5: optional KeySceneInput Input (go.tag = "json:\"input\"", agw.key = "input") // input
    6: optional KeySceneOutput Output (go.tag = "json:\"output\"", agw.key = "output") // output
    7: optional i64 Duration (go.tag = "json:\"duration\"", agw.key = "duration") // It's in milliseconds.
    8: optional i64 StartTime (go.tag = "json:\"start_time\"", api.body = "start_time") // Start time, used for sorting, in milliseconds
    9: optional list<KeyScene> SubKeyScenes (go.tag = "json:\"sub_key_scenes\"", agw.key = "sub_key_scenes") // subscene
}

struct KeySceneInput {
    1: optional string Role (go.tag = "json:\"role\"", agw.key = "role")
    2: optional list<TraceSummaryContent> contentList (go.tag = "json:\"content_list\"", agw.key = "content_list")
}

struct KeySceneOutput {
    1: optional string Role (go.tag = "json:\"role\"", agw.key = "role")
    2: optional list<TraceSummaryContent> contentList (go.tag = "json:\"content_list\"", agw.key = "content_list")
}

struct TraceSummaryContent {
    1: optional string Key (go.tag = "json:\"key\"", agw.key = "key")   // key
    2: optional string Content (go.tag = "json:\"content\"", agw.key = "content") // content
}

struct MessageItem {
    1: optional string Role (go.tag = "json:\"role\"", agw.key = "role")   // role
    2: optional string Content (go.tag = "json:\"content\"", agw.key = "content") // content
}

struct SpanSummary {
    1: optional list<FrontendTag> Tags (go.tag = "json:\"tags\"", agw.key = "tags")
}


struct FrontendTag {
    1: required string Key (go.tag = "json:\"key\"", agw.key = "key")
    2: optional string KeyAlias (go.tag = "json:\"key_alias\"", agw.key = "key_alias") // Multilingual, if there is no configuration value, use the key
    3: required TagType TagType (go.tag = "json:\"tag_type\"", agw.key = "tag_type")
    4: optional Value Value (go.tag = "json:\"value\"", agw.key = "value")
    5: optional FrontedTagType FrontendTagType (go.tag = "json:\"frontend_tag_type\"", agw.key = "frontend_tag_type") // Front-end type for front-end processing
    6: optional bool canCopy (go.tag = "json:\"can_copy\"", agw.key = "can_copy") // Can it be copied?
}


struct TraceSummary{
    1: optional string System (go.tag = "json:\"system\"", agw.key = "system") // System 1 text
    2: optional list<MessageItem> HistoryMessages (go.tag = "json:\"history_messages\"", agw.key = "history_messages") // Level 1 chat history
    3: optional list<KeyScene> KeyScenes (go.tag = "json:\"key_scenes\"", agw.key = "key_scenes")
    4: optional string Input (go.tag = "json:\"input\"", agw.key = "input") // input
    5: optional string Output (go.tag = "json:\"output\"", agw.key = "output") // output
    6: optional i64 Duration (go.tag = "json:\"duration\"", agw.key = "duration") // The duration of the current conversation, in milliseconds
    7: optional string UserID (go.tag = "json:\"user_id\"", agw.key = "user_id") // user ID
}

struct TraceHeader {
    1: optional i64 Duration (go.tag = "json:\"duration\"", agw.key = "duration") // It's in milliseconds.
    2: optional i32 Tokens (agw.key = "tokens")   // Enter the number of tokens consumed
    3: optional i32 StatusCode (go.tag = "json:\"status_code\"", agw.key = "status_code")
    4: optional list<FrontendTag> Tags (go.tag = "json:\"tags\"", agw.key = "tags")
    5: optional string MessageID (go.tag = "json:\"message_id\"", agw.key = "message_id") // Message ID
    6: optional i64 StartTime (go.tag = "json:\"start_time\"", agw.key = "start_time") // It's in milliseconds.
}


struct TraceFrontend {
    1: optional list<TraceFrontendSpan> spans (go.tag = "json:\"spans\"", api.body = "spans")
    2: optional TraceHeader header (go.tag = "json:\"header\"", agw.key = "header")
}

struct TraceFrontendDoubaoCiciDebug {
    1: optional list<TraceFrontendSpan> spans (go.tag = "json:\"spans\"", api.body = "spans")
    2: optional TraceHeader header (go.tag = "json:\"header\"", agw.key = "header")
    3: optional TraceSummary summary (go.tag = "json:\"summary\"", agw.key = "summary")
}
enum InputOutputType {
    TEXT = 0 // Text type
}

struct SpanInputOutput {
    1: optional InputOutputType Type (go.tag = "json:\"type\"", agw.key = "type") // TEXT
    2: optional string Content (go.tag = "json:\"content\"", agw.key = "content")
}

struct TraceFrontendSpan {
    1: string TraceID (go.tag = "json:\"trace_id\"", agw.key = "trace_id")
    2: string LogID (go.tag = "json:\"log_id\"", agw.key = "log_id")
    3: string SpanID (go.tag = "json:\"span_id\"", agw.key = "span_id")
    4: string Type (go.tag = "json:\"type\"", agw.key = "type")
    5: string Name (go.tag = "json:\"name\"", agw.key = "name")
    6: string AliasName (go.tag = "json:\"alias_name\"", agw.key = "alias_name")
    7: string ParentID (go.tag = "json:\"parent_id\"", agw.key = "parent_id")
    8: i64 Duration (go.tag = "json:\"duration\"", agw.key = "duration") // It's in milliseconds.
    9: i64 StartTime (go.tag = "json:\"start_time\"", agw.key = "start_time") // It's in milliseconds.
    10: i32 StatusCode (go.tag = "json:\"status_code\"", agw.key = "status_code")
    11: optional list<TraceTag> Tags (go.tag = "json:\"tags\"", agw.key = "tags")
    12: optional SpanSummary summary (go.tag = "json:\"summary\"", agw.key = "summary") // node details
    13: optional SpanInputOutput Input (go.tag = "json:\"input\"", agw.key = "input")
    14: optional SpanInputOutput Output (go.tag = "json:\"output\"", agw.key = "output")
    15: optional bool IsEntry (go.tag = "json:\"is_entry\"", agw.key = "is_entry") // Is it an entry node?
    16: optional string ProductLine (go.tag = "json:\"product_line\"", agw.key = "product_line") // product line
    17: optional bool IsKeySpan (go.tag = "json:\"is_key_span\"", agw.key = "is_key_span") // Is it a key node?
    18: optional list<string> OwnerList (go.tag = "json:\"owner_list\"", agw.key = "owner_list") // Node owner list, mailbox prefix
    19: optional string RundownDocURL (go.tag = "json:\"rundown_doc_url\"", agw.key = "rundown_doc_url") // Node Details Document
}
