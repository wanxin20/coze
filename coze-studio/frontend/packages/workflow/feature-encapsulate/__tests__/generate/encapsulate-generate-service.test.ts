/*
 * Copyright 2025 coze-dev Authors
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { describe, beforeEach, it, expect } from 'vitest';
import {
  WorkflowDocument,
  type WorkflowNodeEntity,
} from '@flowgram-adapter/free-layout-editor';

import { complexMock } from '../workflow.mock';
import { createContainer } from '../create-container';
import {
  EncapsulateGenerateService,
  type GenerateSubWorkflowNodeOptions,
} from '../../src/generate';

describe('encapsulate-generate-service', () => {
  let encapsulateGenerateService: EncapsulateGenerateService;
  let workflowDocument: WorkflowDocument;
  beforeEach(async () => {
    const container = createContainer();
    workflowDocument = container.get<WorkflowDocument>(WorkflowDocument);
    encapsulateGenerateService = container.get<EncapsulateGenerateService>(
      EncapsulateGenerateService,
    );

    await workflowDocument.fromJSON(complexMock);
  });

  it('should generate workflow json', async () => {
    const nodes = ['102906', '154702'].map(id =>
      workflowDocument.getNode(id),
    ) as WorkflowNodeEntity[];
    const res = await encapsulateGenerateService.generateWorkflowJSON(nodes);
    expect(res).toMatchSnapshot();
  });

  it('should generate sub workflow node', async () => {
    const options: GenerateSubWorkflowNodeOptions = {
      workflowId: '1',
      name: 'test',
      desc: 'test',
      spaceId: '1',
    };
    const res = await encapsulateGenerateService.generateSubWorkflowNode(
      options,
    );
    expect(res).toMatchSnapshot();
  });
});
