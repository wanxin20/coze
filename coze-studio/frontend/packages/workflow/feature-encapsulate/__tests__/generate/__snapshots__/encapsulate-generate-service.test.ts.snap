// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`encapsulate-generate-service > should generate sub workflow node 1`] = `
{
  "data": {
    "inputs": {
      "spaceId": "1",
      "workflowId": "1",
      "workflowVersion": "v0.0.1",
    },
    "nodeMeta": {
      "description": "test",
      "icon": undefined,
      "isImageflow": false,
      "title": "test",
    },
  },
}
`;

exports[`encapsulate-generate-service > should generate workflow json 1`] = `
{
  "edges": [
    {
      "sourceNodeID": "154702",
      "targetNodeID": "102906",
    },
    {
      "sourceNodeID": "100001",
      "targetNodeID": "154702",
    },
    {
      "sourceNodeID": "102906",
      "targetNodeID": "900001",
    },
  ],
  "nodes": [
    {
      "data": {
        "nodeMeta": {},
        "outputs": [],
      },
      "id": "100001",
      "meta": {
        "position": {
          "x": 0,
          "y": 0,
        },
      },
      "type": "1",
    },
    {
      "data": {
        "inputs": {
          "inputParameters": [],
          "terminatePlan": "returnVariables",
        },
        "nodeMeta": {},
      },
      "id": "900001",
      "meta": {
        "position": {
          "x": 1000,
          "y": 0,
        },
      },
      "type": "2",
    },
    {
      "data": {},
      "id": "154702",
      "meta": {
        "position": {
          "x": 918.2411574431025,
          "y": 109.52852376445134,
        },
      },
      "type": "3",
    },
    {
      "data": {
        "inputs": {
          "spaceId": "test_space_id",
          "workflowId": "test_workflow_id",
        },
      },
      "id": "102906",
      "meta": {
        "position": {
          "x": 779.2423264779079,
          "y": -161.54447093414998,
        },
      },
      "type": "9",
    },
  ],
}
`;
